src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java

```diff
@@ -139,102 +146,281 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
+                      new KeyValueBean("lead", 1L),
+                      new KeyValueBean("streams", 3L),
+                      new KeyValueBean("to", 1L),
+                      new KeyValueBean("world", 3L));
+      final List<KeyValueBean>
+              all = fetchRangeOfValues(allRequest,
+              allValues);
+      assertThat(all, equalTo(allValues));
+    
+      // Fetch a range of key-value pairs from the word-count store
+      final List<KeyValueBean> expectedRange = Arrays.asList(
+              new KeyValueBean("all", 1L),
+              new KeyValueBean("hello", 2L),
+              new KeyValueBean("kafka", 2L));
+    
+      final Invocation.Builder
+              request =
+              client.target(baseUrl + "/keyvalues/word-count/range/all/kafka")
+                      .request(MediaType.APPLICATION_JSON_TYPE);
+      final List<KeyValueBean>
+              range = fetchRangeOfValues(request, expectedRange);
+    
+      assertThat(range, equalTo(expectedRange));
+    
+      // Find the instance of the Kafka Streams application that would have the key hello
+      final HostStoreInfo
+              hostWithHelloKey =
+              client.target(baseUrl + "/instance/word-count/hello")
+                      .request(MediaType.APPLICATION_JSON_TYPE).get(HostStoreInfo.class);
+    
+      // Fetch the value for the key hello from the instance.
+      final KeyValueBean result = client.target("http://" + hostWithHelloKey.getHost() +
+              ":" + hostWithHelloKey.getPort() +
+              "/state/keyvalue/word-count/hello")
+              .request(MediaType.APPLICATION_JSON_TYPE).get(KeyValueBean.class);
+    
+      assertThat(result, equalTo(new KeyValueBean("hello", 2L)));
+    
+      // fetch windowed values for a key
+      final List<KeyValueBean>
+              windowedResult =
+              client.target(baseUrl + "/windowed/windowed-word-count/streams/0/" + System
+                      .currentTimeMillis())
+                      .request(MediaType.APPLICATION_JSON_TYPE)
+                    
+                      .get(new GenericType<List<KeyValueBean>>() {
+                      });
+      assertThat(windowedResult.size(), equalTo(1));
+      final KeyValueBean keyValueBean = windowedResult.get(0);
+      assertTrue(keyValueBean.getKey().startsWith("streams"));
+      assertThat(keyValueBean.getValue(), equalTo(3L));
+    }
+  }
+  
+  private void bindHostAndStartRestProxy(int port, String host) {
+    int count = 0;
+    int maxTries = 3;
+    while (count <= maxTries) {
+      try {
+        // Starts the Rest Service on the provided host:port
+        proxy = WordCountInteractiveQueriesExample.startRestProxy(kafkaStreams, port, host);
+      } catch (Exception ex) {
+        log.error(ex.toString());
+      }
+      count++;
+    }
+  }
+  
+  @Test
+  public void shouldStartRestApiOnAnyValidHost() throws Exception {
+    final int port = randomFreeLocalPort();
```

Comment:
Doesn't using `randomFreeLocalPort()` introduce a race condition? The port might be used by some other process until it's used below?

Reply:
Yes @mjsax, it may introduce. And, Therefore, the race condition has already been taken care of using a try/catch block while starting the server.
You can find it here:
WordCountInteractiveQueriesRestService#start(final int port)